### PR TITLE
Test url ensuring inner links always ends with target file instead of directories

### PR DESCRIPTION
**Pull request by arthurscchan**
Create new test to check if inner url links in the fuzz_report does point to static file instead of directory which avoid browser transfer.

### CODE REVIEW

**Comment by arthurscchan** [comment]
Aim to solve #442

**Review by DavidKorczynski — CHANGES_REQUESTED** [review]
I like this overall -- left a few thoughts

File: src/test/test_url.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+import os
+import sys
+import configparser
+
+sys.path.append(os.path.dirname(os.path.realpath(__file__)) + "/../")
+
+from fuzz_introspector import commands  # noqa: E402
+
+
+def is_valid_link(link):
```
[the remark this answers is not in the document]
Comment (arthurscchan) [reply src/test/test_url.py:26]: Done

File: src/test/test_url.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        # Temporary handling on fuzz_report.html only
+        html_file = open("fuzz_report.html", "r")
+        html = html_file.read()
+        html_file.close()
+
+        for link in extract_link_from_html(html):
+            assert is_valid_link(link)
+
+        # Loop and test on HTML result
+#        for file in os.scandir('.'):
```
[the remark this answers is not in the document]
Comment (arthurscchan) [reply src/test/test_url.py:102]: These code exists as a future extension if additional files needed to be tested.

File: src/test/test_url.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    if link.startswith('#') or link.startswith('?'):
+        return True
+
+    # Skip out of scope linkage
+    if link.startswith('http'):
+        if not link.startswith('https://storage.googleapis.com/'):
+            return True
+
+#    if link.startswith('https://storage.googleapis.com/'):
+    file_name = link.split('/')[-1].split('#')[0].split('?')[0]
```
[the remark this answers is not in the document]
Comment (arthurscchan) [reply src/test/test_url.py:37]: I think in general it should be OK to remove all labels (#) and parameters (?) attached to the link and retrieve the last part of the link.

File: src/test/test_url.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    else:
+        return True
+
+
+def extract_link_from_html(html):
+    links = re.findall('href="([^\"]*)"', html)
+    return links
+
+
+def test_run_analysis_on_dir():
```
[the remark this answers is not in the document]
Comment (arthurscchan) [reply src/test/test_url.py:49]: Added new regression testing

**Comment by arthurscchan** [comment]
Regression testing for #427 added.

**Review by DavidKorczynski — CHANGES_REQUESTED** [review]
Could we make the `src/test/TestReport` into a folder `src/test/data/TestReport1`? Thus we can add more test data later on in the `/src/test/data` folder. wdyt?

**Review by DavidKorczynski — APPROVED** [review]

**Comment by DavidKorczynski** [comment]
@arthurscchan I just realised that something went wrong here. I don't think any of the links in the report generated by this test contains `https://storage.googleapis.com/` because this URL is not generated for local runs because the coverage url to `run_analysis_on_dir` is not set to "storage.googleapis..." This line needs to change https://github.com/ossf/fuzz-introspector/blob/2481630390ff5499b3a9e953d71fd46fe978932d/src/test/test_url.py#L91

Could you please check that this test fails if you undo the changes from this PR? https://github.com/ossf/fuzz-introspector/pull/435 This test should fail if those changes are not present.

**Comment by arthurscchan** [comment]
Currently, there is a new subfolder created inside the test folder, named as TestReport.
Inside Test report, there is a config.properties file and a bunch of subfolders.

Each subfolders represent one fuzzer project result which includes their FuzzerLogFile.data and FuzzerLogFile.data.yaml

The config.properties sets the properties for each of the project required this unit testing. with the base folder of each project and specific language specified for future extension.

**Comment by arthurscchan** [comment]
Currently, the unit testing only handle the fuzz_report.html file, commented code in the unit testing file showing how it could expand to search all files ends with .html, or even extending to js or css file in a later stage.

**Comment by arthurscchan** [comment]
Currently, the unit testing with dig up all links in the html which is encapsulated in a href tag.

Links start with '#' '?' are ignored because it points to the current file.
Links start with 'http' and not start with "https://storage.googleapis.com/" are also ignored as they are out of scope link.
Other links, including relative links are being checked one by one to ensure it is pointing to a file instead of a directories.

**Pull request merged**